content/en/blog/_posts/2025-01-15-devsecops-in-action-k8s-secrets/index.md

```diff
@@ -0,0 +1,628 @@
+
+	var isExist, ok bool
+	var envVal string
+
+	if envVal, ok = os.LookupEnv(envVar); ok {
+		isExist = true
+	}
+
+	switch r.URL.Path {
+	case "/":
+		if isExist {
+			w.WriteHeader(http.StatusOK)
+			fmt.Fprintln(w, envVal)
+		} else {
+			w.WriteHeader(http.StatusNotFound)
+			fmt.Fprintln(w, "Secret Not Found")
+		}
+	case "/readiness":
+		if isExist {
+			w.WriteHeader(http.StatusOK)
+			fmt.Fprintln(w, "OK")
+		} else {
+			w.WriteHeader(http.StatusNotFound)
+			fmt.Fprintln(w, "Environment variable wasn't defined")
+		}
+	default:
+		w.WriteHeader(http.StatusNotFound)
+		fmt.Fprintln(w, "404 Not Found")
+	}
+}
+```
+
+## Stage-01: Reject environment vars for storing sensitive data
+
+According to OWASP recommendation, it's necessary to use file instead of env.
+That's why it's necessary to make changes to implementation readiness probes
+and deployment manifests.
+
+[`cmd/main.go`](assets/stage-01/cmd/main.go)
+```go
+const (
+	envVar = "DEMO_SECRET__PASSWD_FILE"
+)
+
+<skipped>
+
+func fileExist(path string) bool {
+	info, err := os.Stat(path)
+	if err == nil {
+		return !info.IsDir()
+	}
+	if errors.Is(err, os.ErrNotExist) {
+		return false
+	}
+	return false
+}
+
+func httpHandle(w http.ResponseWriter, r *http.Request) {
+
+<skipped>
+
+	if envVal, ok = os.LookupEnv(envVar); ok {
+		if fileExist(envVal) {
+			isExist = true
+		}
+	}
+}
+```
+
+Now applications check `DEMO_SECRET__PASSWD_FILE` was defined then application
+checks whether the file exists by value from the environment variable.
+
+[`kustomization.yaml`](assets/stage-01/kustomization/overlays/demo/kustomization.yml)
+```yaml
+
+<skipped>
+
+spec:
+  containers:
+  - envFrom:
+    - secretRef:
+         name: demo-secret
+    volumeMounts:
+    - mountPath: /secrets
+      name: demo-secret
+      readOnly: true
+  volumes:
+  - name: demo-secret
+    secret:
+      optional: true
+      secretName: demo-secret
+
+<skipped>
+```
+
+`spec.containers.0.envFrom` defines all of the Secret's data as container
+environment variables and `spec.volume/spec.containers.0.volumeMounts`
+mounts the Secret as a volume to the Pod.
+
+Excellent! Part of recommendations have been completed, but for the full picture
+it's necessary to solve the following issues:
+
+* The secret file is mounted as a volume and it's necessary to remove after it usage
+(e.g reading)
+* Kind Secret is insecure that's why it's necessary to minimize its use
+* Additional implementation of environment variable such as:
+`DEMO_SECRET__PASSWD_FILE` can Impede applications update.
+
+## Stage-02: Injection of secret to Pod instead of mountpoint
+
+In order to get rid of Secret as mountpoint it's necessary to use Injection of
+Secret via file to Pod. Often secret manager tools perform these functions.
+Also they provide functions such as: dynamic secrets, database credential
+rotation, automated PKI infrastructure, data encryption and tokenization,
+identity-based access, key management, etc.
```

Comment:
I'm not sure this is true. What about environment variables?

Reply:
In the example (stage-03 or viper) is used environment variables in order to define path to file where is stored sensitive data. Some secret managers (agent) can injected file with sensitive data instead of mounting to Pod (e.g. via annotations). Next, application readed file and removed. (also agent can removed it after some delay)  That's why environment variable(s) should store only path to file but not sensitive data.